src/FSharpx.Async/Observable.fs

```diff
@@ -252,73 +252,39 @@ module Observable =
       /// Behaves like AwaitObservable, but calls the specified guarding function
       /// after a subscriber is registered with the observable.
       static member GuardedAwaitObservable (ev1:IObservable<'T1>) guardFunction =
-          let removeObj : IDisposable option ref = ref None
-          let removeLock = new obj()
-          let setRemover r = 
-              lock removeLock (fun () -> removeObj := Some r)
-          let remove() =
-              lock removeLock (fun () ->
-                  match !removeObj with
-                  | Some d -> removeObj := None
-                              d.Dispose()
-                  | None   -> ())
-          synchronize (fun f ->
-          let workflow =
-              Async.FromContinuations((fun (cont,econt,ccont) ->
-                  let rec finish cont value =
-                      remove()
-                      f (fun () -> cont value)
-                  setRemover <|
-                      ev1.Subscribe
-                          ({ new IObserver<_> with
-                              member x.OnNext(v) = finish cont v
-                              member x.OnError(e) = finish econt e
-                              member x.OnCompleted() =
-                                  let msg = "Cancelling the workflow, because the Observable awaited using AwaitObservable has completed."
-                                  finish ccont (new System.OperationCanceledException(msg)) })
-                  guardFunction() ))
           async {
-              let! cToken = Async.CancellationToken
-              let token : CancellationToken = cToken
-              use registration = token.Register((fun _ -> remove()), null)
-              return! workflow
-          })
+              let! token = Async.CancellationToken // capture the current cancellation token
+              return! Async.FromContinuations(fun (cont, econt, ccont) ->
+                  // start a new mailbox processor which will await the result
+                  Agent.Start((fun (mailbox : Agent<Choice<'T1, exn, OperationCanceledException>>) ->
```

Comment:
This one-shot `MailboxProcessor` clearly show a missing abstraction in F# `Async`s - stand alone channel.